linkerd/app/core/src/proxy/server.rs

```diff
@@ -173,41 +183,44 @@ where
-                    .http1_only(true)
-                    .serve_connection(io, HyperServerSvc::new(svc))
-                    .with_upgrades();
-                Either::A(
-                    drain
-                        .watch(conn, |conn| conn.graceful_shutdown())
-                        .map(|_| ())
-                        .map_err(Into::into),
-                )
-            }
-
-            HttpVersion::H2 => {
-                let exec = tokio::executor::DefaultExecutor::current().instrument(info_span!("h2"));
-                let conn = builder
-                    .with_executor(exec)
-                    .http2_only(true)
-                    .http2_initial_stream_window_size(initial_stream_window_size)
-                    .http2_initial_connection_window_size(initial_conn_window_size)
-                    .serve_connection(io, HyperServerSvc::new(http_svc));
-                Either::B(
-                    drain
-                        .watch(conn, |conn| conn.graceful_shutdown())
-                        .map(|_| ())
-                        .map_err(Into::into),
-                )
-            }
-        })
+        Box::new(futures::future::ok::<Self::Response, Error>(Box::new(
+            match http_version {
+                HttpVersion::Http1 => {
+                    // Enable support for HTTP upgrades (CONNECT and websockets).
+                    let svc = upgrade::Service::new(http_svc, drain.clone());
+                    let exec =
+                        tokio::executor::DefaultExecutor::current().instrument(info_span!("http1"));
+                    let conn = builder
+                        .with_executor(exec)
+                        .http1_only(true)
+                        .serve_connection(io, HyperServerSvc::new(svc))
+                        .with_upgrades();
+                    Either::A(
+                        drain
+                            .watch(conn, |conn| conn.graceful_shutdown())
+                            .map(|_| ())
+                            .map_err(Into::into),
+                    )
+                }
+
+                HttpVersion::H2 => {
+                    let exec =
+                        tokio::executor::DefaultExecutor::current().instrument(info_span!("h2"));
+                    let conn = builder
+                        .with_executor(exec)
+                        .http2_only(true)
+                        .http2_initial_stream_window_size(initial_stream_window_size)
+                        .http2_initial_connection_window_size(initial_conn_window_size)
+                        .serve_connection(io, HyperServerSvc::new(http_svc));
+                    Either::B(
+                        drain
+                            .watch(conn, |conn| conn.graceful_shutdown())
+                            .map(|_| ())
+                            .map_err(Into::into),
+                    )
+                }
+            },
+        )))
```

Comment:
I noticed that, since we're boxing the future returned by this match expression, I don't think we need to use `Either` here. I think we can simplify this to (although this was already the case before this PR.)
```suggestion
        Box::new(futures::future::ok::<Self::Response, Error>(
            match http_version {
                HttpVersion::Http1 => {
                    // Enable support for HTTP upgrades (CONNECT and websockets).
                    let svc = upgrade::Service::new(http_svc, drain.clone());
                    let exec =
                        tokio::executor::DefaultExecutor::current().instrument(info_span!("http1"));
                    let conn = builder
                        .with_executor(exec)
                        .http1_only(true)
                        .serve_connection(io, HyperServerSvc::new(svc))
                        .with_upgrades();
                    Box::new(
                        drain
                            .watch(conn, |conn| conn.graceful_shutdown())
                            .map(|_| ())
                            .map_err(Into::into),
                    )
                }

                HttpVersion::H2 => {
                    let exec =
                        tokio::executor::DefaultExecutor::current().instrument(info_span!("h2"));
                    let conn = builder
                        .with_executor(exec)
                        .http2_only(true)
                        .http2_initial_stream_window_size(initial_stream_window_size)
                        .http2_initial_connection_window_size(initial_conn_window_size)
                        .serve_connection(io, HyperServerSvc::new(http_svc));
                    Box::new(
                        drain
                            .watch(conn, |conn| conn.graceful_shutdown())
                            .map(|_| ())
                            .map_err(Into::into),
                    )
                }
            },
        ))
```
Note that this was already the case before this PR, but I only just noticed it. :)

Reply:
Maybe i am way off here, but does that not introduce one additional allocation in case we hit this codepath, or I am failing to understand how that works?

Reply:
Nope, there are the same number of boxes in the returned future. In the current code, we return a boxed `FutureResponse` with a boxed `Either` in it, while in the suggested code we return a boxed `FutureResponse` with a boxed future for each match arm. If that makes sense? There's the same number of boxes.

Reply:
Got it!. Makes sense. Thanks a lot !